pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,504 @@
+    }
+  }
+
+  /**
+   * Can be overridden to provide custom controller Rest API.
+   */
+  @Nullable
+  private ControllerRestApi getControllerRestApi() {
+    return _pushLocations != null ? new DefaultControllerRestApi(_pushLocations, _rawTableName) : null;
+  }
+
+  private void setDirPermission(Path dirPath)
+      throws IOException {
+    if (_defaultPermissionsMask != null) {
+      FsPermission permission = FsPermission.getDirDefault().applyUMask(new FsPermission(_defaultPermissionsMask));
+      _logger.info("Setting permission: {} to directory: {}", permission, dirPath);
+      _fileSystem.setPermission(dirPath, permission);
+    }
+  }
+
+  private void fetchPartitioningConfig() {
+    // Fetch partition info from table config.
+    SegmentPartitionConfig segmentPartitionConfig = _tableConfig.getIndexingConfig().getSegmentPartitionConfig();
+    if (segmentPartitionConfig != null) {
+      Map<String, ColumnPartitionConfig> columnPartitionMap = segmentPartitionConfig.getColumnPartitionMap();
+      assert columnPartitionMap.size() <= 1;
+      if (columnPartitionMap.size() == 1) {
+        _partitionColumn = columnPartitionMap.keySet().iterator().next();
+        _numberOfPartitions = segmentPartitionConfig.getNumPartitions(_partitionColumn);
+        _partitionFunction = segmentPartitionConfig.getFunctionName(_partitionColumn);
+      }
+    } else {
+      _logger.info("Segment partition config is null for table: {}", _tableConfig.getTableName());
+    }
+
+    // Check whether configs overwritten by Hadoop properties.
+    if (_properties.getProperty(PARTITION_COLUMN) != null) {
+      _logger.info("Overriding partition column to: {}", _properties.getProperty(PARTITION_COLUMN));
+      _partitionColumn = _properties.getProperty(PARTITION_COLUMN);
+    }
+    if (_properties.getProperty(NUMBER_OF_PARTITIONS) != null) {
+      _logger.info("Overriding partition number to: {}", _properties.getProperty(NUMBER_OF_PARTITIONS));
+      _numberOfPartitions = Integer.parseInt(_properties.getProperty(NUMBER_OF_PARTITIONS));
+    }
+  }
+
+  private void fetchSortingConfig() {
+    // Fetch sorting info from table config.
+    IndexingConfig indexingConfig = _tableConfig.getIndexingConfig();
+    List<String> sortedColumns = indexingConfig.getSortedColumn();
+    assert sortedColumns.size() <= 1;
+    if (sortedColumns.size() == 1) {
+      _sortedColumn = sortedColumns.get(0);
+    }
+    // Check whether configs overwritten by Hadoop properties.
+    if (_properties.getProperty(SORTED_COLUMN) != null) {
+      _logger.info("Overriding sorted column to: {}", _properties.getProperty(SORTED_COLUMN));
+      _sortedColumn = _properties.getProperty(SORTED_COLUMN);
+    }
+  }
+
+  private void fetchResizingConfig() {
+    if (_properties.getProperty(MINIMUM_NUMBER_OF_OUTPUT_FILES) != null) {
+      _numberOfOutputFiles = Integer.parseInt(_properties.getProperty(MINIMUM_NUMBER_OF_OUTPUT_FILES));
```

Comment:
does min mean we always use this number for output files? it seems like a misleading name in that case, either the variable or the param

Reply:
Yes, we'll use this param to specify the number of reducers if partitioning is disabled. Each reducer will generate at least 1 output file.

Reply:
can we rename it to make it clearer?